### ISSUE TITLE
Feature request: mode that disables Qed checking

### ISSUE THREAD

**Issue by tchajed** [coq/coq#10036]
I'd like a flag to Coq, say `-no-qed-checks`, that disables the type checking at Qed time. We'd use this during development to speed up builds and then turn it back on every so often to double-check the proofs, or perhaps always check proofs for continuous integration.

Our current development spends 35% of compile time processing Qeds.

**Comment by ejgallego** [comment]
I would be happy to give it a go but I'd need to get my proof-save-rework tree in first, as it cleans indeed the proof save path a lot.

A few comments:
- you could implement this at the STM-level, here I am not sure what would the best way is, maybe @gares can provide some pointers.
- another choice is indeed to implement this on lemmas; with current master a good point to start would be to always calling the admitted case here https://github.com/coq/coq/blob/213b5419136e4639f345e171c086b154c14aa62c/vernac/vernacentries.ml#L620 YMMV of course, there will be some tricky corners [due to sections and whatnot] but you can see what you get. Code is a bit duplicated underneath this call, so in this case I'd recommend waiting for the commit doing the unduplication to arrive.

**Comment by ejgallego** [comment]
Also the bug fixed by #9638 may create a lot of trouble here :( :(

**Comment by ejgallego** [comment]
Note that the above method is equivalent to changing the `Qed` by `Admitted` maybe that is a better way to do it ? [A flag in the parser]

**Comment by tchajed** [comment]
Replacing Qed with Admitted is insufficient - I only want to disable checking and do want to make sure that the proof is complete (that is, there are no subgoals). Actually, I always wanted Admitted to refuse to admit proofs with remaining subgoals, but that's a separate issue.

**Comment by ejgallego** [comment]
Add a check in `save_proof_admitted` that will check that there are no open proofs when the flag is on.

**Comment by ejgallego** [comment]
> Add a check in `save_proof_admitted` that will check that there are no open proofs when the flag is on.

In fact you could add it as a separate PR, however I am not sure how useful it'd be in general. The check is pretty simple, using pstate and `Proof.is_complete` I think [the API is a bit confusing still].

**Comment by ejgallego** [comment]
IMHO it is important to prototype something quickly and see what kind of roadblocks you get; once it works for you the implementation can be more polished. As I mentioned doing this in a systematic way may tricky due to sections and who knows what kind of other stuff.

**Comment by JasonGross** [comment]
What about replacing `Qed` with `Unshelve. Grab Existential Variables. Fail idtac. Admitted.`?  (Note that `idtac` fails (with something like "no remaining subgoal") iff the proof is compete.)

**Comment by tchajed** [comment]
@JasonGross yup, that's what I was thinking of hacking together in the absence of changes to Coq. What I was worried about with this hack is that it breaks error message locations, but other than that it's a pretty good solution.

I'll try hijacking `save_proof_admitted` and see what I can put together.

**Comment by tchajed** [comment]
I did some basic tests and this plan needs a bit more work.

Just replacing Qed with Admitted everywhere breaks proofs in sections: whereas the Qed proof had section variables depending on the proof, the Admitted proof uses no section variables. There are a few possibilities I can envision for fixing this:
- Only skip Qed outside a section or when there's an explicit using statement. In our code we'd happily add all the necessary using statement for the performance boost.
- Use the completed proof to construct an appropriate signature for the admitted theorem, but use the `save_proof_admitted` code path.
- Change the `save_proof_proved` path to optionally skip the expensive parts. I don't actually understand what those are (specifically, where and why is the term type checked?).

I like the first the best if I can figure out how to write the appropriate checks.

**Comment by gares** [comment]
> Just replacing Qed with Admitted

`Proof using vars` can make Admitted work in sections. See also `Set Suggest Proof Using`.

The quick hack could be to turn kernel/term_typing.ml in a no-op. You should be able to craft a `j` by hand
here https://github.com/coq/coq/blob/master/kernel/term_typing.ml#L133 instead of calling `infer` and `jusdge_of_cast`. Infer typechecks the proof term, and the type inferred is the compared to the declared type.

**Comment by SkySkimmer** [comment]
>Just replacing Qed with Admitted everywhere breaks proofs in sections: whereas the Qed proof had section variables depending on the proof, the Admitted proof uses no section variables.

That's strange, when I run 
~~~coq
Section S.
  Variables (a:nat) (b : bool).

  Lemma foo : nat.
  Proof. exact a. Admitted.
End S.
Check foo.
~~~
I get `foo : nat -> nat` ie it uses the variable used in the proof and no others. Do you have `Unset Keep Admitted Variables.` somewhere in your script maybe?

**Comment by tchajed** [comment]
Hmm, you're right, it's not that simple - some types are changing and losing parameters, but I'll have to dig into which Admitted is causing it. One issue is that by putting this in Coq itself the standard library has to tolerate the change, including the Prelude, even though I have no issue with running the Qeds in the standard library.

**Comment by tchajed** [comment]
I finally just implemented a sed hack to solve our specific problem: https://github.com/mit-pdos/armada/commit/126f3f687d8c6cb759a7ec3e1762bf0ef4012b0c.

In CI we now selectively replace `Qed` with `Unhselve. Grab Existential Variables. Fail idtac. Admitted` before compiling, getting a huge performance improvement (actually, one of our Qeds is 10 minutes and that's about all we need). As a result, it's a lot less attractive for me personally to fix this in Coq and handle all the edge cases, although I would use the Coq feature if it existed.

**Comment by Zimmi48** [comment]
Side remark @tchajed: `Grab Existential Variables` is supposed to be strictly stronger than `Unshelve` (in principle the two should do the same, but `Grab Existential Variables` will find even the evars that have escaped the shelf because of a bug)..
